File: packages/browser/test/golden.test.ts
// The TypeScript half of the cross-language agreement.
//
// `docs/PLAN.md` Phase 2 says the contract is not trustworthy until every
// maintained language encodes each golden vector to identical bytes. The
// vectors are built in Rust and written to `golden/vectors.json`. This file
// builds the same values in TypeScript and compares.
//
// Every vector name here must appear in the file, and every name in the file
// must appear here. A vector that only one language knows about proves nothing.

import assert from "node:assert/strict";
import { existsSync, readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { test } from "node:test";

import * as collector from "../src/collector-api.ts";
import * as ingest from "../../../generated/typescript/tallyowl-ingest-api/codec.gen.ts";
import type * as ingestTypes from "../../../generated/typescript/tallyowl-ingest-api/types.gen.ts";
import * as control from "../../../generated/typescript/tallyowl-control-api/codec.gen.ts";
import type * as controlTypes from "../../../generated/typescript/tallyowl-control-api/types.gen.ts";

import { bool, bytes, decimal, float, int, measurement, nullValue, property, text, uint, write }
  from "../src/value.ts";

const at = 1_785_628_800_000;

const hex = (u: Uint8Array): string =>
  Array.from(u)
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");

const id = (fill: number): Uint8Array => new Uint8Array(16).fill(fill);

/**
 * Find `golden/vectors.json` by walking up from this file.
 *
 * The compiled output sits at a different depth from the source, so a fixed
 * number of `..` steps would work in one of the two and not the other.
 */
function readGolden(): { name: string; type: string; package: string; bytes: string }[] {
  let directory = dirname(fileURLToPath(import.meta.url));
  for (let step = 0; step < 10; step++) {
    const candidate = join(directory, "golden", "vectors.json");
    if (existsSync(candidate)) {
      return JSON.parse(readFileSync(candidate, "utf8")).vectors;
    }
    directory = dirname(directory);
  }
  throw new Error("golden/vectors.json is not above this file. Run `./tools.sh gen-golden`.");
}

function minimalEnvelope(): collector.Envelope {
  return {
    eventId: id(1),
    kind: "event",
    schemaVersion: 1,
    occurredAt: at,
    sdkName: "tallyowl-driver-rust",
    sdkVersion: "0.0.0",
    properties: [],
  };
}

function fullEnvelope(): collector.Envelope {
  return {
    ...minimalEnvelope(),
    observedAt: at + 1,
    receivedAt: at + 2,
    workspaceId: id(8),
    projectId: id(9),
    sourceId: id(7),
    sequence: 42,
    release: "2026.8.1",
    serviceName: "checkout",
    requestId: "r-1",
    sessionId: "s-1",
    endUserId: "u-1",
    anonymousId: "a-1",
    traceId: id(3),
    spanId: new Uint8Array(8).fill(4),
    consent: { marketing: "denied", analytics: "granted", policyVersion: "2026-01" },
    properties: [
      property("region", text("us-west2"), "collector"),
      property("attempts", uint(3), "driver"),
      property("ratio", float(0.5), "client"),
    ],
    measurements: [measurement("render", float(12.5), "ms")],
  };
}

/** Every vector this language produces, by name. */
function built(): Map<string, Uint8Array> {
  const out = new Map<string, Uint8Array>();
  const typed = (name: string, value: Parameters<typeof write>[0]) =>
    out.set(name, collector.toTypedValueCbor(write(value)));

  typed("typed-value-null", nullValue());
  typed("typed-value-bool", bool(true));
  typed("typed-value-int", int(-3));
  typed("typed-value-uint", uint(3));
  typed("typed-value-float", float(0.5));
  typed("typed-value-decimal", decimal("19.99"));
  typed("typed-value-text", text("us-west2"));
  typed("typed-value-bytes", bytes(new Uint8Array([0xde, 0xad, 0xbe, 0xef])));

  out.set(
    "property-collector-origin",
    collector.toPropertyCbor(property("region", text("us-west2"), "collector")),
  );
  out.set(
    "measurement-decimal-with-unit",
    collector.toMeasurementCbor(measurement("amount", decimal("-0.01"), "USD")),
  );

  out.set("envelope-minimal", collector.toEnvelopeCbor(minimalEnvelope()));
  out.set("envelope-full", collector.toEnvelopeCbor(fullEnvelope()));

  out.set(
    "item-event",
    collector.toTelemetryItemCbor({
      envelope: minimalEnvelope(),
      event: { name: "checkout-started", route: "/checkout" },
    }),
  );

  out.set(
    "item-page-view-with-campaign",
    collector.toTelemetryItemCbor({
      envelope: { ...minimalEnvelope(), kind: "page-view" },
      pageView: {
        route: "/pricing",
        pageTitle: "Pricing",
        campaign: { source: "newsletter", medium: "email", campaign: "spring" },
      },
    }),
  );

  const money = decimal("19.99");
  out.set(
    "item-conversion-exact-money",
    collector.toTelemetryItemCbor({
      envelope: { ...minimalEnvelope(), kind: "conversion" },
      conversion: {
        goal: "purchase",
        value: money.kind === "decimal" ? money.value : undefined,
        currency: "USD",
      },
    }),
  );

  out.set(
    "item-error-with-frames",
    collector.toTelemetryItemCbor({
      envelope: { ...minimalEnvelope(), kind: "error" },
      error: {
        errorType: "TypeError",
        message: "x is not a function",
        handled: false,
        severity: "fatal",
        runtime: "node",
        frames: [{ module: "checkout", function: "submit", line: 42, inApp: true }],
      },
    }),
  );

  out.set(
    "item-session-heartbeat-no-payload",
    collector.toTelemetryItemCbor({
      envelope: { ...minimalEnvelope(), kind: "session-heartbeat" },
    }),
  );

  out.set(
    "batch-two-items",
    collector.toBatchCbor({
      batchId: id(2),
      items: [
        { envelope: minimalEnvelope(), event: { name: "a" } },
        { envelope: minimalEnvelope(), event: { name: "b" } },
      ],
      sealedAt: at,
      compression: "zstd",
    }),
  );

  out.set(
    "commit-batch-response-with-rejection",
    collector.toCommitBatchResponseCbor({
      batchId: id(2),
      accepted: 1,
      committedAt: at,
      satisfiedPolicy: "local-one",
      commitWatermark: 7,
      protocolVersion: 1,
      projectorVersion: 1,
      rejected: [
        {
          eventId: id(5),
          code: "invalid-argument",
          message: "This item carries no project.",
        },
      ],
      deduplicated: false,
    }),
  );

  out.set(
    "service-error-retryable",
    collector.toServiceErrorCbor({
      code: "unavailable",
      message: "We could not reach the durable store.",
      retryable: true,
    }),
  );

  const ingestEnvelope: ingestTypes.Envelope = {
    eventId: id(1),
    kind: "event",
    schemaVersion: 1,
    occurredAt: at,
    sdkName: "tallyowl-browser",
    sdkVersion: "0.0.0",
    properties: [],
  };
  out.set(
    "capture-request-one-event",
    ingest.toCaptureRequestCbor({
      items: [{ envelope: ingestEnvelope, event: { name: "checkout-started", route: "/checkout" } }],
    }),
  );

  const range: controlTypes.TimeRange = {
    rangeStart: at,
    rangeEnd: at + 3_600_000,
    basis: "occurred_at",
  };
  const scanBox: controlTypes.QueryNodeBox = {
    node: "scan",
    scan: { scan: "events", projectId: id(9), range },
  };
  out.set("query-node-scan", control.toQueryNodeBoxCbor(scanBox));

  const literal = write(text("/pricing"));
  out.set(
    "expression-compare",
    control.toExpressionNodeCbor({
      expression: "compare",
      compare: {
        compare: "eq",
        left: control.toExpressionNodeCbor({
          expression: "field",
          field: { name: "route", valueType: "text" },
        }),
        right: control.toExpressionNodeCbor({
          expression: "literal",
          literal: { kind: literal.kind, textValue: literal.textValue },
        }),
      },
    }),
  );

  out.set(
    "query-request-trend",
    control.toQueryRequestCbor({
      algebraVersion: 1,
      consistency: "committed",
      allowPartial: false,
      form: "node",
      node: control.toQueryNodeBoxCbor({
        node: "aggregate",
        aggregate: {
          dimensions: [],
          measures: [{ kind: "count", alias: "events" }],
          interval: { fixedMs: 60_000 },
          input: control.toQueryNodeBoxCbor(scanBox),
        },
      }),
    }),
  );

  return out;
}

test("every golden vector encodes to the same bytes", () => {
  const golden = readGolden();
  const made = built();
  assert.ok(golden.length > 0, "golden/vectors.json holds no vectors");

  const problems: string[] = [];
  for (const vector of golden) {
    const bytes = made.get(vector.name);
    if (bytes === undefined) {
      problems.push(`this language does not build the vector \`${vector.name}\``);
      continue;
    }
    const got = hex(bytes);
    if (got !== vector.bytes) {
      problems.push(
        `\`${vector.name}\` (${vector.type}):\n  Rust:       ${vector.bytes}\n  TypeScript: ${got}`,
      );
    }
    made.delete(vector.name);
  }
  for (const name of made.keys()) {
    problems.push(`this language builds \`${name}\` and the golden file does not hold it`);
  }
  assert.equal(problems.length, 0, problems.join("\n"));
});

File: packages/driver-go/golden_test.go
package tallyowl

// The Go half of the cross-language agreement.
//
// `docs/PLAN.md` Phase 2 says the contract is not trustworthy until every
// maintained language encodes each golden vector to identical bytes. The
// vectors are built in Rust and written to `golden/vectors.json`. This file
// builds the same values in Go and compares.
//
// Every vector name here must appear in the file, and every name in the file
// must appear here. A vector that only one language knows about proves nothing.

import (
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"testing"

	collector "github.com/CatalystCommunity/tallyowl/generated/go/tallyowl-collector-api"
	control "github.com/CatalystCommunity/tallyowl/generated/go/tallyowl-control-api"
	ingest "github.com/CatalystCommunity/tallyowl/generated/go/tallyowl-ingest-api"
)

const at int64 = 1_785_628_800_000

type goldenFile struct {
	Vectors []goldenVector `json:"vectors"`
}

type goldenVector struct {
	Name    string `json:"name"`
	Type    string `json:"type"`
	Package string `json:"package"`
	Bytes   string `json:"bytes"`
}

func readGolden(t *testing.T) goldenFile {
	t.Helper()
	path := filepath.Join("..", "..", "golden", "vectors.json")
	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("golden/vectors.json could not be read: %v", err)
	}
	var out goldenFile
	if err := json.Unmarshal(raw, &out); err != nil {
		t.Fatalf("golden/vectors.json could not be parsed: %v", err)
	}
	return out
}

func id(fill byte) []byte {
	out := make([]byte, 16)
	for i := range out {
		out[i] = fill
	}
	return out
}

func text(s string) *string { return &s }

func minimalEnvelope() collector.Envelope {
	return collector.Envelope{
		EventId:       id(1),
		Kind:          "event",
		SchemaVersion: 1,
		OccurredAt:    collector.Timestamp(at),
		SdkName:       "tallyowl-driver-rust",
		SdkVersion:    "0.0.0",
		Properties:    collector.PropertyList{},
	}
}

func fullEnvelope(t *testing.T) collector.Envelope {
	t.Helper()
	env := minimalEnvelope()
	observed := collector.Timestamp(at + 1)
	received := collector.Timestamp(at + 2)
	sequence := uint64(42)
	session := collector.SessionId("s-1")
	trace := collector.TraceId(id(3))
	span := collector.SpanId([]byte{4, 4, 4, 4, 4, 4, 4, 4})
	workspace := collector.WorkspaceId(id(8))
	project := collector.ProjectId(id(9))
	source := collector.SourceId(id(7))

	env.ObservedAt = &observed
	env.ReceivedAt = &received
	env.WorkspaceId = &workspace
	env.ProjectId = &project
	env.SourceId = &source
	env.Sequence = &sequence
	env.Release = text("2026.8.1")
	env.ServiceName = text("checkout")
	env.RequestId = text("r-1")
	env.SessionId = &session
	env.EndUserId = text("u-1")
	env.AnonymousId = text("a-1")
	env.TraceId = &trace
	env.SpanId = &span
	env.Consent = &collector.Consent{
		Marketing:     "denied",
		Analytics:     "granted",
		PolicyVersion: text("2026-01"),
	}
	env.Properties = collector.PropertyList{
		Property("region", Text("us-west2"), "collector"),
		Property("attempts", Uint(3), "driver"),
		Property("ratio", Float(0.5), "client"),
	}
	render, err := Measurement("render", Float(12.5), "ms")
	if err != nil {
		t.Fatalf("a float is a number: %v", err)
	}
	measurements := collector.MeasurementList{render}
	env.Measurements = &measurements
	return env
}

// built returns every vector this language produces, by name.
func built(t *testing.T) map[string][]byte {
	t.Helper()
	out := map[string][]byte{}

	typed := func(name string, v Value) {
		out[name] = collector.EncodeTypedValue(v.Wire())
	}
	typed("typed-value-null", Null())
	typed("typed-value-bool", Bool(true))
	typed("typed-value-int", Int(-3))
	typed("typed-value-uint", Uint(3))
	typed("typed-value-float", Float(0.5))
	typed("typed-value-decimal", MustDecimal("19.99"))
	typed("typed-value-text", Text("us-west2"))
	typed("typed-value-bytes", Bytes([]byte{0xde, 0xad, 0xbe, 0xef}))

	out["property-collector-origin"] = collector.EncodeProperty(
		Property("region", Text("us-west2"), "collector"))

	amount, err := Measurement("amount", MustDecimal("-0.01"), "USD")
	if err != nil {
		t.Fatalf("a decimal is a number: %v", err)
	}
	out["measurement-decimal-with-unit"] = collector.EncodeMeasurement(amount)

	out["envelope-minimal"] = collector.EncodeEnvelope(minimalEnvelope())
	out["envelope-full"] = collector.EncodeEnvelope(fullEnvelope(t))

	event := collector.EventPayload{Name: "checkout-started", Route: text("/checkout")}
	out["item-event"] = collector.EncodeTelemetryItem(collector.TelemetryItem{
		Envelope: minimalEnvelope(), Event: &event,
	})

	pageEnvelope := minimalEnvelope()
	pageEnvelope.Kind = "page-view"
	pageView := collector.PageViewPayload{
		Route:     "/pricing",
		PageTitle: text("Pricing"),
		Campaign: &collector.CampaignParameters{
			Source:   text("newsletter"),
			Medium:   text("email"),
			Campaign: text("spring"),
		},
	}
	out["item-page-view-with-campaign"] = collector.EncodeTelemetryItem(collector.TelemetryItem{
		Envelope: pageEnvelope, PageView: &pageView,
	})

	conversionEnvelope := minimalEnvelope()
	conversionEnvelope.Kind = "conversion"
	money := MustDecimal("19.99").Decimal
	conversion := collector.ConversionPayload{
		Goal: "purchase", Value: &money, Currency: text("USD"),
	}
	out["item-conversion-exact-money"] = collector.EncodeTelemetryItem(collector.TelemetryItem{
		Envelope: conversionEnvelope, Conversion: &conversion,
	})

	errorEnvelope := minimalEnvelope()
	errorEnvelope.Kind = "error"
	line := uint64(42)
	errorPayload := collector.ErrorPayload{
		ErrorType: "TypeError",
		Message:   "x is not a function",
		Handled:   false,
		Severity:  "fatal",
		Runtime:   text("node"),
		Frames: []collector.StackFrame{{
			Module:   text("checkout"),
			Function: text("submit"),
			Line:     &line,
			InApp:    true,
		}},
	}
	out["item-error-with-frames"] = collector.EncodeTelemetryItem(collector.TelemetryItem{
		Envelope: errorEnvelope, Error: &errorPayload,
	})

	heartbeatEnvelope := minimalEnvelope()
	heartbeatEnvelope.Kind = "session-heartbeat"
	out["item-session-heartbeat-no-payload"] = collector.EncodeTelemetryItem(
		collector.TelemetryItem{Envelope: heartbeatEnvelope})

	a := collector.EventPayload{Name: "a"}
	b := collector.EventPayload{Name: "b"}
	compression := collector.Compression("zstd")
	out["batch-two-items"] = collector.EncodeBatch(collector.Batch{
		BatchId: id(2),
		Items: []collector.TelemetryItem{
			{Envelope: minimalEnvelope(), Event: &a},
			{Envelope: minimalEnvelope(), Event: &b},
		},
		SealedAt:    collector.Timestamp(at),
		Compression: &compression,
	})

	deduplicated := false
	out["commit-batch-response-with-rejection"] = collector.EncodeCommitBatchResponse(
		collector.CommitBatchResponse{
			BatchId:          id(2),
			Accepted:         1,
			CommittedAt:      collector.Timestamp(at),
			SatisfiedPolicy:  "local-one",
			CommitWatermark:  7,
			ProtocolVersion:  1,
			ProjectorVersion: 1,
			Rejected: []collector.RejectedItem{{
				EventId: id(5),
				Code:    "invalid-argument",
				Message: "This item carries no project.",
			}},
			Deduplicated: &deduplicated,
		})

	out["service-error-retryable"] = collector.EncodeServiceError(collector.ServiceError{
		Code:      "unavailable",
		Message:   "We could not reach the durable store.",
		Retryable: true,
	})

	ingestEvent := ingest.EventPayload{Name: "checkout-started", Route: text("/checkout")}
	out["capture-request-one-event"] = ingest.EncodeCaptureRequest(ingest.CaptureRequest{
		Items: []ingest.TelemetryItem{{
			Envelope: ingest.Envelope{
				EventId:       id(1),
				Kind:          "event",
				SchemaVersion: 1,
				OccurredAt:    ingest.Timestamp(at),
				SdkName:       "tallyowl-browser",
				SdkVersion:    "0.0.0",
				Properties:    ingest.PropertyList{},
			},
			Event: &ingestEvent,
		}},
	})

	timeRange := control.TimeRange{
		RangeStart: control.Timestamp(at),
		RangeEnd:   control.Timestamp(at + 3_600_000),
		Basis:      "occurred_at",
	}
	scan := control.ScanNode{Scan: "events", ProjectId: id(9), Range: timeRange}
	scanBox := control.QueryNodeBox{Node: "scan", Scan: &scan}
	out["query-node-scan"] = control.EncodeQueryNodeBox(scanBox)

	fieldRef := control.FieldRef{Name: "route", ValueType: text("text")}
	fieldNode := control.ExpressionNode{Expression: "field", Field: &fieldRef}
	literalValue := Text("/pricing").Wire()
	literalNode := control.ExpressionNode{
		Expression: "literal",
		Literal: &control.TypedValue{
			Kind:      control.TypedValueKind(literalValue.Kind),
			TextValue: literalValue.TextValue,
		},
	}
	compare := control.CompareExpr{
		Compare: "eq",
		Left:    control.EncodeExpressionNode(fieldNode),
		Right:   control.EncodeExpressionNode(literalNode),
	}
	out["expression-compare"] = control.EncodeExpressionNode(
		control.ExpressionNode{Expression: "compare", Compare: &compare})

	interval := control.Interval{FixedMs: func() *control.DurationMs {
		v := control.DurationMs(60_000)
		return &v
	}()}
	aggregate := control.AggregateNode{
		Dimensions: []control.Dimension{},
		Measures:   []control.Measure{{Kind: "count", Alias: "events"}},
		Interval:   &interval,
		Input:      control.EncodeQueryNodeBox(scanBox),
	}
	aggregateBox := control.QueryNodeBox{Node: "aggregate", Aggregate: &aggregate}
	node := control.QueryNodeRef(control.EncodeQueryNodeBox(aggregateBox))
	out["query-request-trend"] = control.EncodeQueryRequest(control.QueryRequest{
		AlgebraVersion: 1,
		Consistency:    "committed",
		AllowPartial:   false,
		Form:           "node",
		Node:           &node,
	})

	return out
}

func TestEveryGoldenVectorEncodesToTheSameBytes(t *testing.T) {
	golden := readGolden(t)
	made := built(t)

	if len(golden.Vectors) == 0 {
		t.Fatal("golden/vectors.json holds no vectors")
	}

	for _, vector := range golden.Vectors {
		bytes, ok := made[vector.Name]
		if !ok {
			t.Errorf("this language does not build the vector `%s`", vector.Name)
			continue
		}
		if got := hex.EncodeToString(bytes); got != vector.Bytes {
			t.Errorf("`%s` (%s):\n  Rust: %s\n  Go:   %s",
				vector.Name, vector.Type, vector.Bytes, got)
		}
		delete(made, vector.Name)
	}

	for name := range made {
		t.Errorf("this language builds `%s` and the golden file does not hold it", name)
	}
}

func TestEveryValueKindSurvivesAWriteAndARead(t *testing.T) {
	values := []Value{
		Null(), Bool(true), Int(-3), Uint(9), Float(0.5),
		MustDecimal("19.99"), Text("us-west2"), Bytes([]byte{1, 2, 3}),
	}
	for _, value := range values {
		back, err := ReadValue(value.Wire())
		if err != nil {
			t.Fatalf("%s did not read back: %v", value.Kind, err)
		}
		if back.Kind != value.Kind || back.String() != value.String() {
			t.Errorf("%s came back as %s (%q against %q)",
				value.Kind, back.Kind, back.String(), value.String())
		}
	}
}

func TestAnUnsignedValueStaysUnsigned(t *testing.T) {
	// This is the distinction the earlier bare choice could not carry in a
	// dynamically typed language. It is the reason the shape changed.
	wire := Uint(9).Wire()
	if wire.Kind != "uint" {
		t.Fatalf("an unsigned value travels as %q", wire.Kind)
	}
	if wire.IntValue != nil {
		t.Error("an unsigned value carried a signed number as well")
	}
	if wire.UintValue == nil || *wire.UintValue != 9 {
		t.Error("an unsigned value did not carry its number")
	}
}

func TestAKindThatNamesAnAbsentValueIsRefused(t *testing.T) {
	wire := Text("x").Wire()
	wire.TextValue = nil
	if _, err := ReadValue(wire); err == nil {
		t.Error("a value that says it holds text and carries none was accepted")
	}
}

func TestMoneyKeepsItsExactDigits(t *testing.T) {
	for _, want := range []string{"19.99", "-0.01", "0.1", "1000", "-12345.6789"} {
		value := MustDecimal(want)
		if got := value.String(); got != want {
			t.Errorf("%q came back as %q", want, got)
		}
	}
	if _, err := Decimal("not a number"); err == nil {
		t.Error("text that is not a number was accepted as one")
	}
}

func TestAMeasurementRefusesAValueThatIsNotANumber(t *testing.T) {
	if _, err := Measurement("duration", Float(1.5), "ms"); err != nil {
		t.Errorf("a float is a number: %v", err)
	}
	if _, err := Measurement("name", Text("x"), ""); err == nil {
		t.Error("text was accepted as a measurement")
	}
}
